src/Formats/JSONUtils.h

```diff
@@ -134,6 +143,106 @@ namespace JSONUtils
 
     bool skipUntilFieldInObject(ReadBuffer & in, const String & desired_field_name, const FormatSettings::JSON & settings);
     void skipTheRestOfObject(ReadBuffer & in, const FormatSettings::JSON & settings);
+
+    template <typename ReturnType>
+    ReturnType deserializeEmpyStringAsDefaultOrNested(IColumn & column, ReadBuffer & istr, const FormatSettings & settings, auto && deserializer)
```

Comment:
Actually, I would expected something similar to `SerializationNullable::deserializeNullAsDefaultOrNestedText*` that can be used both in `JSONUtins::readField` and in nested types serializations. Don't quite understand why we implement logic with nulls as defaults here when it can be hidden inside `deserializer`.
I would do smth like this (very similar to `deserializeTextEscapedAndRawImpl` in `SerializationNullable`:
```c++
template <typename ReturnType, typename NestedDeserialize>
ReturnType deserializeEmpyStringAsDefaultOrNested(IColumn & column, ReadBuffer & istr, NestedDeserialize && deserialize_nested, bool & is_empty_string)
{
    static constexpr auto throw_exception = std::is_same_v<ReturnType, void>;

    static constexpr auto EMPTY_STRING = "\"\"";
    static constexpr auto EMPTY_STRING_LENGTH = std::string_view(EMPTY_STRING).length();

    if (istr.eof() || *istr.position() != EMPTY_STRING[0])
        return deserialize_nested(column, istr);

    auto do_deserialize = [&](IColumn & column_, ReadBuffer & buf_, auto && check_for_empty_string, auto && deserialize) -> ReturnType
    {
        is_empty_string = check_for_empty_string(buf_);
        if (is_empty_string)
        {
            column_.insertDefault();
        }
        else
        {
            if constexpr (throw_exception)
                deserialize(column_, buf_);
            else
                return deserialize(column_, buf_);
        }

        return ReturnType(true);
    };

    if (istr.available() >= EMPTY_STRING_LENGTH)
    {
        /// We have enough data in buffer to check if we have an empty string.
        auto check_for_empty_string = [](ReadBuffer & buf_)
        {
            auto * pos = buf_.position();
            if (checkString(EMPTY_STRING, buf_))
                return true;
            buf_.position() = pos;
            return false;
        };

        return do_deserialize(column, istr, check_for_empty_string, deserialize_nested);
    }

    /// We don't have enough data in buffer to check if we have an empty string.
    /// Use PeekableReadBuffer to make a checkpoint before checking for an
    /// empty string and rollback if check was failed.
    auto check_for_empty_string = [](ReadBuffer & buf_) -> bool
    {
        auto & peekable_buf = assert_cast<PeekableReadBuffer &>(buf_);
        peekable_buf.setCheckpoint();
        SCOPE_EXIT(peekable_buf.dropCheckpoint());
        if (checkString(EMPTY_STRING, peekable_buf))
            return true;
        peekable_buf.rollbackToCheckpoint();
        return false;
    };

    auto deserialize_nested_with_check = [&deserialize_nested](IColumn & column_, ReadBuffer & buf_) -> ReturnType
    {
        auto & peekable_buf = assert_cast<PeekableReadBuffer &>(buf_);
        if constexpr (throw_exception)
            deserialize_nested(column_, peekable_buf);
        else if (!deserialize_nested(column_, peekable_buf))
            return ReturnType(false);

        if (unlikely(peekable_buf.hasUnreadData()))
            throw Exception(ErrorCodes::LOGICAL_ERROR, "Incorrect state while parsing JSON: PeekableReadBuffer has unread data in own memory: {}", String(peekable_buf.position(), peekable_buf.available()));

        return ReturnType(true);
    };

    PeekableReadBuffer peekable_buf(istr, true);
    return do_deserialize(column, peekable_buf, check_for_empty_string, deserialize_nested_with_check);
}
```
And all logic with `null_as_default` will be hidden inside `deserialize_nested`.

Also we need to move implementation to .cpp file. Probably we will have to change `NestedDeserialize` type to `std::function<ReturnType(IColumn &, ReadBuffer &)>`

Reply:
I did something similar to your idea.